# Análise Exploratória - Vendas em uma Startup

### Kedro

In [1]:
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from kedro.io import DataCatalog
    catalog: DataCatalog

In [2]:
%load_ext kedro.ipython

[09/20/25 01:42:20] INFO     Using 'c:\Users\Zeus\Projetos\Projetos DS\Projetos                     __init__.py:270
                             Preditiva\otimizacao-de-ads-para-startup\.venv\Lib\site-packages\kedro                
                             \framework\project\rich_logging.yml' as logging configuration.                        

                    INFO     Registered line magic '%reload_kedro'                                   __init__.py:63

                    INFO     Registered line magic '%load_node'                                      __init__.py:65

                    INFO     Resolved project path as: c:\Users\Zeus\Projetos\Projetos DS\Projetos  __init__.py:180
                             Preditiva\otimizacao-de-ads-para-startup.                                             
                             To set a different path, run '%reload_kedro <project_root>'                           

[09/20/25 01:42:26] INFO     Kedro is sending anonymous usage data with the sole purpose of improving plugin.py:243
                             the product. No personal data or IP addresses are stored on our side. To              
                             opt out, set the `KEDRO_DISABLE_TELEMETRY` or `DO_NOT_TRACK` environment              
                             variables, or create a `.telemetry` file in the current working                       
                             directory with the contents `consent: false`. To hide this message,                   
                             explicitly grant or deny consent. Read more at                                        
                             https://docs.kedro.org/en/stable/configuration/telemetry.html                         

[09/20/25 01:42:27] INFO     Kedro project Otimizacao de Ads para Startup                           __init__.py:146

                    INFO     Defined global variable 'context', 'session', 'catalog' and            __init__.py:147
                             'pipelines'                                                                           

[09/20/25 01:42:28] INFO     Registered line magic 'run_viz'                                        __init__.py:153

### Importações de Bibliotecas

In [3]:

# Importando as bibliotecas necessárias
import warnings

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns


In [4]:

# Configurações
warnings.filterwarnings("ignore")
plt.style.use("seaborn-v0_8")
sns.set_palette("husl")
plt.rcParams["figure.figsize"] = (12, 8)
plt.rcParams["font.size"] = 12

# Configurando para exibir todas as colunas
pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)


## Visão inicial das bases disponíveis

#### Metadados

In [5]:
df_campanha_metadados = catalog.load("metadados")
df_campanha_metadados

[09/20/25 01:42:32] INFO     Loading data from metadados (ExcelDataset)...                     data_catalog.py:1046

Variável  \
0               id_do_anuncio   
1                    campanha   
2                faixa_etária   
3                        sexo   
4      categoria_de_interesse   
5           qte_de_impressões   
6               qte_de_clicks   
7  valor_investido_no_anúncio   
8   qte_de_Vendas_após_Clique   

                                           Descrição  
0  Identificação única do anúncio mostrado na tim...  
1  Campanha realizada. Uma campanha é uma combina...  
2  Faixa de idade do lead em que o anúncio foi mo...  
3        Sexo do lead em que o anúncio foi mostrado.  
4  Identificação da categoria de interesse do lea...  
5  Quantidade de visualizações que este anúncio o...  
6     Quantidade de cliques que este anúncio obteve.  
7  Uma das formas de custo do Facebook Ads é o cu...  
8  Quantidade de vendas de cursos obtidos após o ...

In [6]:
# Aumentar o limite de largura da coluna para melhor visualização
pd.set_option("display.max_colwidth", None)
df_campanha_metadados

Variável  \
0               id_do_anuncio   
1                    campanha   
2                faixa_etária   
3                        sexo   
4      categoria_de_interesse   
5           qte_de_impressões   
6               qte_de_clicks   
7  valor_investido_no_anúncio   
8   qte_de_Vendas_após_Clique   

                                                                                                                                              Descrição
0                                                                                      Identificação única do anúncio mostrado na timeline do Facebook.
1                                                       Campanha realizada. Uma campanha é uma combinação de públicos alvo para se mostrar os anúncios.
2                                                                                                 Faixa de idade do lead em que o anúncio foi mostrado.
3                                                                                                           Sexo do lead em que o anúncio foi mostrado.
4                                                                        Identificação da categoria de interesse do lead em que o anúncio foi mostrado.
5                                                                                                  Quantidade de visualizações que este anúncio obteve.
6                                                                                                        Quantidade de cliques que este anúncio obteve.
7  Uma das formas de custo do Facebook Ads é o custo por clique. Esta variável mede o valor que o Facebook cobrou pela quantidade de cliques recebidos.
8                                                                                      Quantidade de vendas de cursos obtidos após o clique no anúncio.

In [7]:
# Voltando a visualização padrão
pd.set_option("display.max_colwidth", 50)

#### Campanha

In [8]:
df_campanha = catalog.load("campanhas")
df_campanha.head()

[09/20/25 01:42:33] INFO     Loading data from campanhas (ExcelDataset)...                     data_catalog.py:1046

id_do_anuncio    campanha faixa_etária sexo  categoria_de_interesse  \
0         708746  Campanha A        30-34    M                      15   
1         708749  Campanha A        30-34    M                      16   
2         708771  Campanha A        30-34    M                      20   
3         708815  Campanha A        30-34    M                      28   
4         708818  Campanha A        30-34    M                      28   

   qte_de_impressões  qte_de_clicks  valor_investido_no_anúncio  \
0               7350              1                        1.43   
1              17861              2                        1.82   
2                693              0                        0.00   
3               4259              1                        1.25   
4               4133              1                        1.29   

   Qte_de_Vendas_após_Clique  
0                          3  
1                          2  
2                          0  
3                          1  
4                          2

In [9]:
df_campanha = df_campanha.rename(columns={
    "qte_de_impressões": "impressoes",
    "qte_de_clicks": "cliques",
    "valor_investido_no_anúncio": "custo",
    "Qte_de_Vendas_após_Clique": "vendas"
})

In [10]:
from unidecode import unidecode

df_campanha.columns = [unidecode(col) for col in df_campanha.columns]

In [11]:
df_campanha.head()

id_do_anuncio    campanha faixa_etaria sexo  categoria_de_interesse  \
0         708746  Campanha A        30-34    M                      15   
1         708749  Campanha A        30-34    M                      16   
2         708771  Campanha A        30-34    M                      20   
3         708815  Campanha A        30-34    M                      28   
4         708818  Campanha A        30-34    M                      28   

   impressoes  cliques  custo  vendas  
0        7350        1   1.43       3  
1       17861        2   1.82       2  
2         693        0   0.00       0  
3        4259        1   1.25       1  
4        4133        1   1.29       2

#### Categorias

In [12]:
df_categorias = catalog.load("categorias")
df_categorias.head()

                    INFO     Loading data from categorias (ExcelDataset)...                    data_catalog.py:1046

categoria_de_interesse  descrição_da_categoria
0                       2  Business and Industry 
1                       7                     NaN
2                      10                     NaN
3                      15                     NaN
4                      16          Entertainment

In [13]:
df_categorias.columns = [unidecode(col) for col in df_categorias.columns]
df_categorias

categoria_de_interesse     descricao_da_categoria
0                        2     Business and Industry 
1                        7                        NaN
2                       10                        NaN
3                       15                        NaN
4                       16             Entertainment 
5                       18                        NaN
6                       19                        NaN
7                       20                        NaN
8                       21  Family and relationships 
9                       22                        NaN
10                      23                        NaN
11                      24                        NaN
12                      25      Fitness and wellness 
13                      26                        NaN
14                      27                        NaN
15                      28                        NaN
16                      29            Food and drink 
17                      30                        NaN
18                      31                        NaN
19                      32                        NaN
20                      36    Hobbies and activities 
21                      63                        NaN
22                      64                        NaN
23                      65                        NaN
24                      66      Shopping and fashion 
25                     100                        NaN
26                     101                        NaN
27                     102                        NaN
28                     103       Sports and outdoors 
29                     104                        NaN
30                     105                        NaN
31                     106                        NaN
32                     107                Technology 
33                     108                        NaN
34                     109                        NaN
35                     110                        NaN
36                     111                        NaN
37                     112                        NaN
38                     113                        NaN
39                     114                        NaN

Provável que estivessem usando celulas mescladas no excel, tendo texto apenas na primeira linha da mescla. Um ffill resolve isso.

#### Categoria Detalhada

In [14]:
df_categoria_detalhada = catalog.load("categoria_detalhada")
df_categoria_detalhada.head(20)

                    INFO     Loading data from categoria_detalhada (ExcelDataset)...           data_catalog.py:1046

categoria_detalhada
0                              Business and Industry 
1                                                 NaN
2   Advertising Agriculture Architecture Aviation ...
3                                                 NaN
4   Investment banking Online banking Retail banking 
5                                                 NaN
6                       Business Construction Design 
7                                                 NaN
8      Fashion design Graphic design Interior design 
9                                                 NaN
10  Economics Engineering Entrepreneurship Health ...
11                                                NaN
12                                             Online
13                                 Digital marketing 
14                                                NaN
15                                Display advertising
16                                    Email marketing
17                                 Online advertising
18           Search engine optimization Social media 
19                                                NaN

- *Comportamento estranho das categorias*

A base de categoria parece ser mais abrangente e a Planilha 8 mais detalhada. Porém ambas tem diversos registros nulos e em Planilha8 não há uma referência de número da categoria.

Olhando pela planilha, na aba depara_categoria parece que Business and Industry se aplicaria às próximas 3 linhas também, que poderia ser usado um ffill pra preencher, mas assim teriamos diversos números de categorias de interesse com descrição repetida.

Enquanto se olharmos a planilha 8, dá pra ver que dentro de business industry temos 17 outras quebras, ou um tree map, onde as linhas sequenciadas representam subcategorias, mas que não se encaixaria nas 3 descrições nulas da aba de categorias.

### Decisão de como cruzar as categorias com suas descrições

In [15]:
df_campanha["categoria_de_interesse"] = df_campanha["categoria_de_interesse"].astype("category")
df_campanha[["categoria_de_interesse"]].describe()

categoria_de_interesse
count                     1143
unique                      40
top                         16
freq                       140

In [16]:
df_categoria_detalhada.describe()

categoria_detalhada
count                      148
unique                     148
top     Business and Industry 
freq                         1

In [17]:
df_categorias["descricao_da_categoria"].describe()


count                          9
unique                         9
top       Business and Industry 
freq                           1
Name: descricao_da_categoria, dtype: object

In [18]:
df_categorias["descricao_da_categoria"].ffill().describe()


count              40
unique              9
top       Technology 
freq                8
Name: descricao_da_categoria, dtype: object

Dentre as categoiras presentes nas campanhas, são 40 valores únicos. Nas campanhas detalhadas da planilha8 existem 148 categorias e no df_cateogias são 9, porém realizando o preenchimento ffill ficariam 40 também, igual ao número presente nas campanhas, mas não seriam únicas.

Vou usar a df_categorias com ffill pra dar join com a df_campanhas

In [19]:
df_categorias["descricao_da_categoria"] = df_categorias["descricao_da_categoria"].ffill()

In [20]:
df_campanha_merged = (
    df_campanha.merge(
        df_categorias,
        how="left",
        on="categoria_de_interesse"
    )
)

In [21]:
import sys
sys.path.append("..")
from src.otimizacao_de_ads_para_startup.pipelines.data_processing.nodes import *


In [33]:

df_perfil = process_df_perfil(catalog.load("campanhas"), catalog.load("categorias"))


[09/20/25 01:46:40] INFO     Loading data from campanhas (ExcelDataset)...                     data_catalog.py:1046

                    INFO     Loading data from categorias (ExcelDataset)...                    data_catalog.py:1046

## Análise Descritiva

In [ ]:
df_perfil.to_csv("df_perfil.csv", index=False)

In [34]:
# Ajustar a "categoria_de_interesse" pra categorica
df_campanha_merged["categoria_de_interesse"] = df_campanha_merged["categoria_de_interesse"].astype("category")
df_campanha_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   id_do_anuncio           1143 non-null   int64   
 1   campanha                1143 non-null   object  
 2   faixa_etaria            1143 non-null   object  
 3   sexo                    1143 non-null   object  
 4   categoria_de_interesse  1143 non-null   category
 5   impressoes              1143 non-null   int64   
 6   cliques                 1143 non-null   int64   
 7   custo                   1143 non-null   float64 
 8   vendas                  1143 non-null   int64   
 9   descricao_da_categoria  1143 non-null   object  
 10  perfil                  1143 non-null   object  
 11  faturamento             1143 non-null   int64   
 12  lucro                   1143 non-null   float64 
 13  ctr                     1143 non-null   float64 
 14  tc                      

In [35]:
df_campanha_merged.head()

id_do_anuncio    campanha faixa_etaria sexo categoria_de_interesse  \
0         708746  Campanha A        30-34    M                     15   
1         708749  Campanha A        30-34    M                     16   
2         708771  Campanha A        30-34    M                     20   
3         708815  Campanha A        30-34    M                     28   
4         708818  Campanha A        30-34    M                     28   

   impressoes  cliques  custo  vendas  descricao_da_categoria          perfil  \
0        7350        1   1.43       3  Business and Industry   30-34 | M | 15   
1       17861        2   1.82       2          Entertainment   30-34 | M | 16   
2         693        0   0.00       0          Entertainment   30-34 | M | 20   
3        4259        1   1.25       1   Fitness and wellness   30-34 | M | 28   
4        4133        1   1.29       2   Fitness and wellness   30-34 | M | 28   

   faturamento   lucro       ctr   tc   cpc        cc         roi    conv_i  
0          255  253.57  0.000136  3.0  1.43  0.476667  177.321685  0.000408  
1          170  168.18  0.000112  1.0  0.91  0.910000   92.406592  0.000112  
2            0    0.00  0.000000  NaN   NaN       NaN         NaN  0.000000  
3           85   83.75  0.000235  1.0  1.25  1.250000   67.000000  0.000235  
4          170  168.71  0.000242  2.0  1.29  0.645000  130.782950  0.000484

In [36]:
def full_describe(df) -> None:
    print("Info")
    print(df.info())
    print("\ndescribe de categóricas")
    display(df.describe(include=["object", "category"]))
    print("\ndescribe de numéricas")
    display(df.describe(include="number"))
    print("\nHead")
    display(df.head())

# Verificando valores únicos nas colunas categóricas
def value_counts(df) -> None:

    print("=== VALORES ÚNICOS NAS COLUNAS CATEGÓRICAS ===")

    for coluna in df.select_dtypes(include=["object", "category"]).columns:
        vc = df[coluna].value_counts().reset_index()
        vc.columns = [coluna, "qtd"]
        vc["percentual"] = (vc["qtd"] / vc["qtd"].sum() * 100).round(2)
        display(vc)


In [37]:
full_describe(df_campanha_merged)

Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   id_do_anuncio           1143 non-null   int64   
 1   campanha                1143 non-null   object  
 2   faixa_etaria            1143 non-null   object  
 3   sexo                    1143 non-null   object  
 4   categoria_de_interesse  1143 non-null   category
 5   impressoes              1143 non-null   int64   
 6   cliques                 1143 non-null   int64   
 7   custo                   1143 non-null   float64 
 8   vendas                  1143 non-null   int64   
 9   descricao_da_categoria  1143 non-null   object  
 10  perfil                  1143 non-null   object  
 11  faturamento             1143 non-null   int64   
 12  lucro                   1143 non-null   float64 
 13  ctr                     1143 non-null   float64 
 14  tc                 

campanha faixa_etaria  sexo  categoria_de_interesse  \
count         1143         1143  1143                    1143   
unique           3            4     2                      40   
top     Campanha C        30-34     M                      16   
freq           625          426   592                     140   

       descricao_da_categoria          perfil  
count                    1143            1143  
unique                      9             284  
top            Entertainment   30-34 | M | 16  
freq                      264              24


describe de numéricas


id_do_anuncio    impressoes      cliques        custo       vendas  \
count   1.143000e+03  1.143000e+03  1143.000000  1143.000000  1143.000000   
mean    9.872611e+05  1.867321e+05    33.390201    51.360656     5.956255   
std     1.939928e+05  3.127622e+05    56.892438    86.908418     7.728301   
min     7.087460e+05  8.700000e+01     0.000000     0.000000     0.000000   
25%     7.776325e+05  6.503500e+03     1.000000     1.480000     1.000000   
50%     1.121185e+06  5.150900e+04     8.000000    12.370000     2.000000   
75%     1.121804e+06  2.217690e+05    37.500000    60.025000    10.000000   
max     1.314415e+06  3.052003e+06   421.000000   639.949998    86.000000   

       faturamento        lucro          ctr          tc         cpc  \
count  1143.000000  1143.000000  1143.000000  936.000000  936.000000   
mean    506.281715   454.921059     0.000164    0.736614    1.499347   
std     656.905556   606.338770     0.000115    1.270894    0.232879   
min       0.000000  -332.989999     0.000000    0.000000    0.180000   
25%      85.000000    65.740000     0.000100    0.124649    1.390000   
50%     170.000000   168.620000     0.000160    0.333333    1.498273   
75%     850.000000   805.100000     0.000234    1.000000    1.644364   
max    7310.000000  6670.050002     0.001059   11.000000    2.212000   

               cc         roi       conv_i  
count  936.000000  936.000000  1143.000000  
mean          inf   43.498702     0.000095  
std           NaN   78.371726     0.000147  
min      0.090000   -1.000000     0.000000  
25%      1.590000    5.881630     0.000008  
50%      4.738333   16.938799     0.000040  
75%     12.351753   52.459118     0.000114  
max           inf  943.444408     0.001389


Head


id_do_anuncio    campanha faixa_etaria sexo categoria_de_interesse  \
0         708746  Campanha A        30-34    M                     15   
1         708749  Campanha A        30-34    M                     16   
2         708771  Campanha A        30-34    M                     20   
3         708815  Campanha A        30-34    M                     28   
4         708818  Campanha A        30-34    M                     28   

   impressoes  cliques  custo  vendas  descricao_da_categoria          perfil  \
0        7350        1   1.43       3  Business and Industry   30-34 | M | 15   
1       17861        2   1.82       2          Entertainment   30-34 | M | 16   
2         693        0   0.00       0          Entertainment   30-34 | M | 20   
3        4259        1   1.25       1   Fitness and wellness   30-34 | M | 28   
4        4133        1   1.29       2   Fitness and wellness   30-34 | M | 28   

   faturamento   lucro       ctr   tc   cpc        cc         roi    conv_i  
0          255  253.57  0.000136  3.0  1.43  0.476667  177.321685  0.000408  
1          170  168.18  0.000112  1.0  0.91  0.910000   92.406592  0.000112  
2            0    0.00  0.000000  NaN   NaN       NaN         NaN  0.000000  
3           85   83.75  0.000235  1.0  1.25  1.250000   67.000000  0.000235  
4          170  168.71  0.000242  2.0  1.29  0.645000  130.782950  0.000484

In [38]:
value_counts(df_campanha_merged)

=== VALORES ÚNICOS NAS COLUNAS CATEGÓRICAS ===


campanha  qtd  percentual
0  Campanha C  625       54.68
1  Campanha B  464       40.59
2  Campanha A   54        4.72

faixa_etaria  qtd  percentual
0        30-34  426       37.27
1        45-49  259       22.66
2        35-39  248       21.70
3        40-44  210       18.37

sexo  qtd  percentual
0    M  592       51.79
1    F  551       48.21

categoria_de_interesse  qtd  percentual
0                      16  140       12.25
1                      10   85        7.44
2                      29   77        6.74
3                      27   60        5.25
4                      15   51        4.46
5                      28   51        4.46
6                      20   49        4.29
7                      64   48        4.20
8                      63   46        4.02
9                      18   43        3.76
10                     26   41        3.59
11                     21   36        3.15
12                     32   33        2.89
13                     22   33        2.89
14                     19   32        2.80
15                     25   26        2.27
16                     31   25        2.19
17                      2   25        2.19
18                     30   25        2.19
19                      7   24        2.10
20                     24   24        2.10
21                     23   23        2.01
22                     36   21        1.84
23                     65   19        1.66
24                     66   11        0.96
25                    107    8        0.70
26                    110    8        0.70
27                    101    7        0.61
28                    112    7        0.61
29                    108    7        0.61
30                    102    7        0.61
31                    105    7        0.61
32                    109    6        0.52
33                    111    6        0.52
34                    113    6        0.52
35                    100    6        0.52
36                    106    5        0.44
37                    104    5        0.44
38                    103    5        0.44
39                    114    5        0.44

descricao_da_categoria  qtd  percentual
0             Entertainment   264       23.10
1     Business and Industry   185       16.19
2      Fitness and wellness   178       15.57
3            Food and drink   160       14.00
4    Hobbies and activities   134       11.72
5  Family and relationships   116       10.15
6                Technology    53        4.64
7      Shopping and fashion    31        2.71
8       Sports and outdoors    22        1.92

perfil  qtd  percentual
0     30-34 | M | 16   24        2.10
1     30-34 | F | 16   23        2.01
2     45-49 | F | 16   20        1.75
3     30-34 | M | 10   17        1.49
4     35-39 | M | 16   17        1.49
..               ...  ...         ...
279  45-49 | F | 109    1        0.09
280  45-49 | F | 110    1        0.09
281  45-49 | F | 111    1        0.09
282  45-49 | F | 113    1        0.09
283  45-49 | F | 114    1        0.09

[284 rows x 3 columns]

Como queremos identificar os melhores perfis pra uma nova campanha, acredito que seja melhor remover a coluna de campanha e id_do_anuncio pra agrupar tudo por perfil (faixa_etária, sexo e descrição_da_categoria).

Mas antes, quero validar se o desempenho dos perfis são parecidos em diferentes campanhas.

In [39]:
df_campanha_merged["perfil"] = (
    df_campanha_merged["faixa_etaria"].astype(str) + " | " +
    df_campanha_merged["sexo"].astype(str) + " | " +
    df_campanha_merged["categoria_de_interesse"].astype(str)
)

Também precisamos de métricas melhores pra fazer as comparações. Usaremos algumas métricas padrões em marketing

In [40]:
# Valor de venda do curso definido na apresentação do problema (disponível no Readme).
VALOR_VENDA = 85

# Calculando os KPIs
df_campanha_merged["faturamento"] = df_campanha_merged["vendas"] * VALOR_VENDA   # Receita total gerada pelas vendas
df_campanha_merged["lucro"] = df_campanha_merged["faturamento"] - df_campanha_merged["custo"]   # Lucro líquido = receita - custo do anúncio
df_campanha_merged["ctr"] = (df_campanha_merged["cliques"] / df_campanha_merged["impressoes"])   # Click Through Rate (%) = % de impressões que geraram clique
df_campanha_merged["tc"] = (df_campanha_merged["vendas"] / df_campanha_merged["cliques"]) # Taxa de conversão (%) = % de cliques que geraram venda
df_campanha_merged["cpc"] = df_campanha_merged["custo"] / df_campanha_merged["cliques"]   # Custo por clique médio
df_campanha_merged["cc"] = df_campanha_merged["custo"] / df_campanha_merged["vendas"]   # Custo por conversão (venda)
df_campanha_merged["roi"] = (df_campanha_merged["faturamento"] - df_campanha_merged["custo"]) / df_campanha_merged["custo"]   # Retorno sobre investimento (%) = lucro / custo * 100
df_campanha_merged["conv_i"] = df_campanha_merged["vendas"] / df_campanha_merged["impressoes"]  # Taxa de conversão por impressão

In [41]:
df_campanha_merged.head()

id_do_anuncio    campanha faixa_etaria sexo categoria_de_interesse  \
0         708746  Campanha A        30-34    M                     15   
1         708749  Campanha A        30-34    M                     16   
2         708771  Campanha A        30-34    M                     20   
3         708815  Campanha A        30-34    M                     28   
4         708818  Campanha A        30-34    M                     28   

   impressoes  cliques  custo  vendas  descricao_da_categoria          perfil  \
0        7350        1   1.43       3  Business and Industry   30-34 | M | 15   
1       17861        2   1.82       2          Entertainment   30-34 | M | 16   
2         693        0   0.00       0          Entertainment   30-34 | M | 20   
3        4259        1   1.25       1   Fitness and wellness   30-34 | M | 28   
4        4133        1   1.29       2   Fitness and wellness   30-34 | M | 28   

   faturamento   lucro       ctr   tc   cpc        cc         roi    conv_i  
0          255  253.57  0.000136  3.0  1.43  0.476667  177.321685  0.000408  
1          170  168.18  0.000112  1.0  0.91  0.910000   92.406592  0.000112  
2            0    0.00  0.000000  NaN   NaN       NaN         NaN  0.000000  
3           85   83.75  0.000235  1.0  1.25  1.250000   67.000000  0.000235  
4          170  168.71  0.000242  2.0  1.29  0.645000  130.782950  0.000484

In [42]:
df_campanha_merged.sort_values(by="perfil", ascending=True).head(20)

id_do_anuncio    campanha faixa_etaria sexo categoria_de_interesse  \
468          951033  Campanha B        30-34    F                     10   
825         1121742  Campanha C        30-34    F                     10   
173          747791  Campanha B        30-34    F                     10   
469          951035  Campanha B        30-34    F                     10   
234          776579  Campanha B        30-34    F                     10   
77           734785  Campanha B        30-34    F                     10   
172          747790  Campanha B        30-34    F                     10   
824         1121741  Campanha C        30-34    F                     10   
1091        1314357  Campanha C        30-34    F                    101   
1092        1314358  Campanha C        30-34    F                    102   
1093        1314359  Campanha C        30-34    F                    103   
1094        1314360  Campanha C        30-34    F                    104   
1095        1314361  Campanha C        30-34    F                    105   
1096        1314362  Campanha C        30-34    F                    106   
1097        1314363  Campanha C        30-34    F                    107   
1098        1314364  Campanha C        30-34    F                    108   
1099        1314365  Campanha C        30-34    F                    109   
1100        1314366  Campanha C        30-34    F                    110   
1101        1314367  Campanha C        30-34    F                    111   
1102        1314368  Campanha C        30-34    F                    112   

      impressoes  cliques       custo  vendas  descricao_da_categoria  \
468         5517        1    1.230000       1  Business and Industry    
825       213016       30   44.220000      13  Business and Industry    
173        31393        8   10.960001       2  Business and Industry    
469         1539        0    0.000000       0  Business and Industry    
234        26910        5    7.230000       1  Business and Industry    
77          5576        1    1.530000       2  Business and Industry    
172         2077        0    0.000000       0  Business and Industry    
824       318042       46   64.410000       9  Business and Industry    
1091      524306       81  113.680003      15   Shopping and fashion    
1092      104496        9   11.430000       7   Shopping and fashion    
1093      452519       68   99.520002       7    Sports and outdoors    
1094      442919       76  110.780002      23    Sports and outdoors    
1095      596831       86  120.879999      11    Sports and outdoors    
1096      173912       26   35.540000       5    Sports and outdoors    
1097      780967       86  119.640002      26             Technology    
1098      132124        8   11.190000       3             Technology    
1099      623137      100  138.920001      12             Technology    
1100       99020       10   14.480000       4             Technology    
1101      665817      117  163.800000      33             Technology    
1102      699232       80  111.989999      15             Technology    

               perfil  faturamento        lucro       ctr        tc       cpc  \
468    30-34 | F | 10           85    83.770000  0.000181  1.000000  1.230000   
825    30-34 | F | 10         1105  1060.780000  0.000141  0.433333  1.474000   
173    30-34 | F | 10          170   159.039999  0.000255  0.250000  1.370000   
469    30-34 | F | 10            0     0.000000  0.000000       NaN       NaN   
234    30-34 | F | 10           85    77.770000  0.000186  0.200000  1.446000   
77     30-34 | F | 10          170   168.470000  0.000179  2.000000  1.530000   
172    30-34 | F | 10            0     0.000000  0.000000       NaN       NaN   
824    30-34 | F | 10          765   700.590000  0.000145  0.195652  1.400217   
1091  30-34 | F | 101         1275  1161.319997  0.000154  0.185185  1.403457   
1092  30-34 | F | 102          595   583.570000  0.000086  0.7777

Podemos identificar acima que existem valores diferentes pra um mesmo perfil dentro de uma mesma campanha. Isso pode acontecer por haver mais de um design usado pro mesmo perfil na mesma campanhas.
Além disso, os desempenhos de cada design tiveram diferenças relevantes. Caso tivessemos uma base que indicasse quais designes foram usados em quais id_de_anuncios, poderíamos explorar, além do perfil de usuário, qual design teve melhor desempenho.

Como não temos tais informações, agruparei o desempenho dos perfis, ignorando o id do anuncio e a campanha.

In [43]:
df_perfil = (
    df_campanha_merged.groupby(
        ["faixa_etaria",
         "sexo",
         "categoria_de_interesse",
         "descricao_da_categoria",
         "perfil"],
         observed=True
    )
    .agg({
        "impressoes": "sum",
        "cliques": "sum",
        "custo": "sum",
        "vendas": "sum"}
    )
    .reset_index()
)

In [44]:
df_perfil["faturamento"] = df_perfil["vendas"] * VALOR_VENDA   # Receita total gerada pelas vendas
df_perfil["lucro"] = df_perfil["faturamento"] - df_perfil["custo"]   # Lucro líquido = receita - custo do anúncio
df_perfil["ctr"] = (df_perfil["cliques"] / df_perfil["impressoes"])   # Click Through Rate (%) = % de impressões que geraram clique
df_perfil["tc"] = (df_perfil["vendas"] / df_perfil["cliques"]) # Taxa de conversão (%) = % de cliques que geraram venda
df_perfil["cpc"] = df_perfil["custo"] / df_perfil["cliques"]   # Custo por clique médio
df_perfil["cc"] = df_perfil["custo"] / df_perfil["vendas"]   # Custo por conversão (venda)
df_perfil["roi"] = (df_perfil["faturamento"] - df_perfil["custo"]) / df_perfil["custo"]   # Retorno sobre investimento (%) = lucro / custo * 100
df_perfil["conversao"] = df_perfil["vendas"] / df_perfil["impressoes"]  # Taxa de conversão por impressão

In [45]:
full_describe(df_perfil)

Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284 entries, 0 to 283
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   faixa_etaria            284 non-null    object  
 1   sexo                    284 non-null    object  
 2   categoria_de_interesse  284 non-null    category
 3   descricao_da_categoria  284 non-null    object  
 4   perfil                  284 non-null    object  
 5   impressoes              284 non-null    int64   
 6   cliques                 284 non-null    int64   
 7   custo                   284 non-null    float64 
 8   vendas                  284 non-null    int64   
 9   faturamento             284 non-null    int64   
 10  lucro                   284 non-null    float64 
 11  ctr                     284 non-null    float64 
 12  tc                      281 non-null    float64 
 13  cpc                     281 non-null    float64 
 14  cc                   

faixa_etaria sexo  categoria_de_interesse descricao_da_categoria  \
count           284  284                     284                    284   
unique            4    2                      40                      9   
top           30-34    F                      15            Technology    
freq             72  144                       8                     53   

               perfil  
count             284  
unique            284  
top     30-34 | F | 2  
freq                1


describe de numéricas


impressoes      cliques        custo      vendas   faturamento  \
count  2.840000e+02   284.000000   284.000000  284.000000    284.000000   
mean   7.515311e+05   134.383803   206.708556   23.971831   2037.605634   
std    1.001406e+06   187.275361   286.107323   27.153261   2308.027145   
min    6.520000e+02     0.000000     0.000000    0.000000      0.000000   
25%    1.287982e+05    20.000000    30.215000    7.750000    658.750000   
50%    3.737660e+05    65.000000   103.120001   13.000000   1105.000000   
75%    8.667928e+05   166.250000   247.794998   32.250000   2741.250000   
max    6.110946e+06  1321.000000  1888.630003  176.000000  14960.000000   

              lucro         ctr          tc         cpc          cc  \
count    284.000000  284.000000  281.000000  281.000000  281.000000   
mean    1830.897078    0.000173    0.535443    1.517036         inf   
std     2136.180396    0.000065    1.099989    0.155152         NaN   
min     -120.899998    0.000000    0.000000    0.655000    0.135455   
25%      513.810001    0.000132    0.100000    1.424706    2.905833   
50%      990.600000    0.000166    0.250000    1.478545    6.203247   
75%     2517.775001    0.000213    0.510638    1.617447   15.208000   
max    14003.090003    0.000439   11.000000    1.996250         inf   

              roi   conversao  
count  281.000000  284.000000  
mean    29.544697    0.000065  
std     65.486467    0.000087  
min     -1.000000    0.000000  
25%      4.589164    0.000019  
50%     12.702502    0.000040  
75%     28.251506    0.000072  
max    626.516774    0.000664


Head


faixa_etaria sexo categoria_de_interesse  descricao_da_categoria  \
0        30-34    F                      2  Business and Industry    
1        30-34    F                      7  Business and Industry    
2        30-34    F                     10  Business and Industry    
3        30-34    F                     15  Business and Industry    
4        30-34    F                     16          Entertainment    

           perfil  impressoes  cliques       custo  vendas  faturamento  \
0   30-34 | F | 2      252951       45   67.200000      11          935   
1   30-34 | F | 7      317046       58   84.860000      12         1020   
2  30-34 | F | 10      604070       91  129.580000      28         2380   
3  30-34 | F | 15     1669161      262  391.600002      46         3910   
4  30-34 | F | 16     2693782      355  529.709996      62         5270   

         lucro       ctr        tc       cpc        cc        roi  conversao  
0   867.800000  0.000178  0.244444  1.493333  6.109091  12.913690   0.000043  
1   935.140000  0.000183  0.206897  1.463103  7.071667  11.019797   0.000038  
2  2250.420000  0.000151  0.307692  1.423956  4.627857  17.367032   0.000046  
3  3518.399998  0.000157  0.175573  1.494656  8.513044   8.984678   0.000028  
4  4740.290004  0.000132  0.174648  1.492141  8.543710   8.948840   0.000023

Temos que nos atentar aos valores infinitos gerados em cc por conta da divisão por zero. Vou substituir por nulo 

In [46]:
df_perfil["cc"] = df_perfil["cc"].replace([np.inf, -np.inf], np.nan)

In [47]:
value_counts(df_perfil)

=== VALORES ÚNICOS NAS COLUNAS CATEGÓRICAS ===


faixa_etaria  qtd  percentual
0        30-34   72       25.35
1        40-44   72       25.35
2        35-39   71       25.00
3        45-49   69       24.30

sexo  qtd  percentual
0    F  144        50.7
1    M  140        49.3

categoria_de_interesse  qtd  percentual
0                      15    8        2.82
1                      10    8        2.82
2                      19    8        2.82
3                      16    8        2.82
4                      20    8        2.82
5                      23    8        2.82
6                      22    8        2.82
7                      21    8        2.82
8                      64    8        2.82
9                      63    8        2.82
10                    107    8        2.82
11                     25    8        2.82
12                     27    8        2.82
13                     26    8        2.82
14                     31    8        2.82
15                     29    8        2.82
16                    110    8        2.82
17                     18    7        2.46
18                      7    7        2.46
19                      2    7        2.46
20                     32    7        2.46
21                     30    7        2.46
22                     28    7        2.46
23                     24    7        2.46
24                     65    7        2.46
25                     36    7        2.46
26                    102    7        2.46
27                    101    7        2.46
28                    112    7        2.46
29                    108    7        2.46
30                    105    7        2.46
31                    100    6        2.11
32                    109    6        2.11
33                    111    6        2.11
34                    113    6        2.11
35                     66    6        2.11
36                    106    5        1.76
37                    104    5        1.76
38                    103    5        1.76
39                    114    5        1.76

descricao_da_categoria  qtd  percentual
0                Technology    53       18.66
1      Fitness and wellness    31       10.92
2             Entertainment    31       10.92
3  Family and relationships    31       10.92
4     Business and Industry    30       10.56
5            Food and drink    30       10.56
6    Hobbies and activities    30       10.56
7      Shopping and fashion    26        9.15
8       Sports and outdoors    22        7.75

perfil  qtd  percentual
0      30-34 | F | 2    1        0.35
1      30-34 | F | 7    1        0.35
2     30-34 | F | 10    1        0.35
3     30-34 | F | 15    1        0.35
4     30-34 | F | 16    1        0.35
..               ...  ...         ...
279  45-49 | M | 109    1        0.35
280  45-49 | M | 110    1        0.35
281  45-49 | M | 112    1        0.35
282  45-49 | M | 113    1        0.35
283  45-49 | M | 114    1        0.35

[284 rows x 3 columns]

In [48]:
df_perfil.head()

faixa_etaria sexo categoria_de_interesse  descricao_da_categoria  \
0        30-34    F                      2  Business and Industry    
1        30-34    F                      7  Business and Industry    
2        30-34    F                     10  Business and Industry    
3        30-34    F                     15  Business and Industry    
4        30-34    F                     16          Entertainment    

           perfil  impressoes  cliques       custo  vendas  faturamento  \
0   30-34 | F | 2      252951       45   67.200000      11          935   
1   30-34 | F | 7      317046       58   84.860000      12         1020   
2  30-34 | F | 10      604070       91  129.580000      28         2380   
3  30-34 | F | 15     1669161      262  391.600002      46         3910   
4  30-34 | F | 16     2693782      355  529.709996      62         5270   

         lucro       ctr        tc       cpc        cc        roi  conversao  
0   867.800000  0.000178  0.244444  1.493333  6.109091  12.913690   0.000043  
1   935.140000  0.000183  0.206897  1.463103  7.071667  11.019797   0.000038  
2  2250.420000  0.000151  0.307692  1.423956  4.627857  17.367032   0.000046  
3  3518.399998  0.000157  0.175573  1.494656  8.513044   8.984678   0.000028  
4  4740.290004  0.000132  0.174648  1.492141  8.543710   8.948840   0.000023

In [52]:

# Identificando os melhores perfis baseado no ROI
melhores_perfis = (
    df_perfil.groupby(
        ["faixa_etaria", "sexo", "descricao_da_categoria"],
        observed=True
        )
        .agg({
            "roi": "mean",
            "custo": "mean",
            "vendas": "mean",
            "ctr": "mean",
            "tc": "mean"
        })
        .round(2)
)

In [54]:

melhores_perfis = melhores_perfis.sort_values("roi", ascending=False)
print("Melhores perfis baseado no ROI:")
display(melhores_perfis.head(10))

Melhores perfis baseado no ROI:


roi   custo  vendas  ctr    tc
faixa_etaria sexo descricao_da_categoria                                    
40-44        M    Sports and outdoors      236.83   15.04   10.33  0.0  4.18
30-34        M    Shopping and fashion     195.32   37.58   16.25  0.0  3.39
35-39        M    Food and drink           163.50  211.89   41.00  0.0  2.59
                  Shopping and fashion     114.10   19.79   13.00  0.0  2.04
40-44        M    Shopping and fashion     104.00   12.15   13.67  0.0  1.84
30-34        M    Technology                94.98   31.02   15.20  0.0  1.73
45-49        M    Sports and outdoors       74.49   18.67   11.00  0.0  1.30
                  Hobbies and activities    68.74   38.75   15.50  0.0  1.28
35-39        M    Technology                55.24   19.82   10.43  0.0  0.97
             F    Food and drink            52.67  123.39   11.50  0.0  0.57

In [56]:
VERBA_TOTAL = 20000

# Sugestão de alocação baseada no ROI
print(f"\n=== SUGESTÃO DE ALOCAÇÃO DA VERBA DE R$ {VERBA_TOTAL:,.2f} ===")

# Alocando mais verba para perfis com maior ROI
top_perfis = melhores_perfis.head(5)
total_roi_top = top_perfis["roi"].sum()

for i, (perfil, dados) in enumerate(top_perfis.iterrows()):
    # Alocação proporcional ao ROI
    alocacao = (dados["roi"] / total_roi_top) * VERBA_TOTAL
    print(f"\nPerfil {i+1}: {perfil}")
    print(f"  ROI médio: {dados['roi']:.2f}%")
    print(f"  Alocação sugerida: R$ {alocacao:,.2f}")
    print(f"  Vendas esperadas: {alocacao / dados['custo'] * dados['vendas']:.0f}")
    print(f"  Lucro esperado: R$ {alocacao / dados['custo'] * dados['vendas'] * VALOR_VENDA - alocacao:,.2f}")

print("\n=== ANÁLISE CONCLUÍDA ===")
print("Principais descobertas e recomendações foram apresentadas acima.")
print("Considere implementar as sugestões de alocação da verba para maximizar o ROI.")



=== SUGESTÃO DE ALOCAÇÃO DA VERBA DE R$ 20,000.00 ===

Perfil 1: ('40-44', 'M', 'Sports and outdoors ')
  ROI médio: 236.83%
  Alocação sugerida: R$ 5,820.71
  Vendas esperadas: 3998
  Lucro esperado: R$ 333,997.87

Perfil 2: ('30-34', 'M', 'Shopping and fashion ')
  ROI médio: 195.32%
  Alocação sugerida: R$ 4,800.49
  Vendas esperadas: 2076
  Lucro esperado: R$ 171,641.20

Perfil 3: ('35-39', 'M', 'Food and drink ')
  ROI médio: 163.50%
  Alocação sugerida: R$ 4,018.43
  Vendas esperadas: 778
  Lucro esperado: R$ 62,073.59

Perfil 4: ('35-39', 'M', 'Shopping and fashion ')
  ROI médio: 114.10%
  Alocação sugerida: R$ 2,804.30
  Vendas esperadas: 1842
  Lucro esperado: R$ 153,777.44

Perfil 5: ('40-44', 'M', 'Shopping and fashion ')
  ROI médio: 104.00%
  Alocação sugerida: R$ 2,556.07
  Vendas esperadas: 2876
  Lucro esperado: R$ 241,890.25

=== ANÁLISE CONCLUÍDA ===
Principais descobertas e recomendações foram apresentadas acima.
Considere implementar as sugestões de alocação da ve

# WIP

Exportação

In [ ]:
# Identificando colunas quantitativas
colunas_quantitativas = df_campanha.select_dtypes(include=[np.number]).columns

# Criando histogramas
fig, axes = plt.subplots(2, 4, figsize=(20, 12))
axes = axes.ravel()

for i, coluna in enumerate(colunas_quantitativas):
    if i < 8:  # Limitando a 8 gráficos
        axes[i].hist(df_campanha[coluna], bins=20, alpha=0.7, edgecolor="black")
        axes[i].set_title(f"Distribuição de {coluna}")
        axes[i].set_xlabel(coluna)
        axes[i].set_ylabel("Frequência")
        axes[i].grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

In [ ]:

print("=== MEDIDAS RESUMO DAS VARIÁVEIS QUANTITATIVAS ===")
medidas_resumo = df_campanha[colunas_quantitativas].describe()
display(medidas_resumo.round(2))

In [ ]:
# Matriz de correlação
correlacao = df_campanha[colunas_quantitativas].corr()

plt.figure(figsize=(12, 10))
sns.heatmap(correlacao, annot=True, cmap="coolwarm", center=0,
            square=True, linewidths=0.5, cbar_kws={"shrink": .8})
plt.title("Matriz de Correlação entre Variáveis Quantitativas")
plt.tight_layout()
plt.show()

In [ ]:

print("\n=== CORRELAÇÕES MAIS FORTES ===")
# Encontrando as correlações mais fortes (excluindo correlação com ela mesma)
correlacoes = []
for i in range(len(correlacao.columns)):
    for j in range(i+1, len(correlacao.columns)):
        correlacoes.append((
            correlacao.columns[i],
            correlacao.columns[j],
            correlacao.iloc[i, j]
        ))

correlacoes_df = pd.DataFrame(correlacoes, columns=["Variável 1", "Variável 2", "Correlação"])
correlacoes_df = correlacoes_df.sort_values("Correlação", key=abs, ascending=False)
display(correlacoes_df.head(15).round(3))


In [ ]:

# Análise de ROI por segmentação
print("=== ANÁLISE DE ROI POR SEGMENTAÇÃO ===")

for coluna_qual in colunas_qualitativas:
    print(f"\nROI por {coluna_qual}:")
    roi_por_grupo = df_campanha.groupby(coluna_qual)["roi"].agg(["mean", "std", "count"]).round(2)
    roi_por_grupo.columns = ["ROI Médio (%)", "Desvio Padrão", "Quantidade"]
    print(roi_por_grupo)

    # Gráfico de boxplot
    plt.figure(figsize=(10, 6))
    df_campanha.boxplot(column="roi", by=coluna_qual, figsize=(10, 6))
    plt.title(f"Distribuição do ROI por {coluna_qual}")
    plt.suptitle("")  # Remove o título automático
    plt.show()


In [ ]:
# Calculando métricas do funil de vendas
print("=== FUNIL DE VENDAS GERAL ===")

total_impressoes = df_campanha["impressoes"].sum()
total_cliques = df_campanha["cliques"].sum()
total_vendas = df_campanha["vendas"].sum()
total_custo = df_campanha["custo"].sum()

ctr_geral = (total_cliques / total_impressoes) * 100
tc_geral = (total_vendas / total_cliques) * 100
roi_geral = ((total_vendas * VALOR_VENDA - total_custo) / total_custo) * 100

print(f"Total de Impressões: {total_impressoes:,}")
print(f"Total de Cliques: {total_cliques:,}")
print(f"Total de Vendas: {total_vendas:,}")
print(f"CTR Geral: {ctr_geral:.2f}%")
print(f"Taxa de Conversão Geral: {tc_geral:.2f}%")
print(f"ROI Geral: {roi_geral:.2f}%")
print(f"Custo Total: R$ {total_custo:,.2f}")
print(f"Faturamento Total: R$ {total_vendas * VALOR_VENDA:,.2f}")
print(f"Lucro Total: R$ {total_vendas * VALOR_VENDA - total_custo:,.2f}")

In [ ]:

# Visualizando o funil de vendas
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6))

# Gráfico de barras do funil
etapas = ["Impressões", "Cliques", "Vendas"]
valores = [total_impressoes, total_cliques, total_vendas]
percentuais = [100, ctr_geral, tc_geral]

ax1.bar(etapas, valores, color=["#FF6B6B", "#4ECDC4", "#45B7D1"])
ax1.set_title("Funil de Vendas - Valores Absolutos")
ax1.set_ylabel("Quantidade")
for i, v in enumerate(valores):
    ax1.text(i, v + max(valores)*0.01, f"{v:,}", ha="center", va="bottom")

# Gráfico de percentuais
ax2.bar(etapas, percentuais, color=["#FF6B6B", "#4ECDC4", "#45B7D1"])
ax2.set_title("Funil de Vendas - Percentuais")
ax2.set_ylabel("Percentual (%)")
for i, v in enumerate(percentuais):
    ax2.text(i, v + 1, f"{v:.2f}%", ha="center", va="bottom")

plt.tight_layout()
plt.show()


## Dados de Pesquisa

### Base Pesquisa

In [ ]:
# Dados da pesquisa de mercado
excel_pesquisa = pd.read_excel(caminho_dados / "dados_pesquisa.xlsx", sheet_name=None)
print("Resumo das abas encontradas:")
for nome, df in excel_pesquisa.items():
    print(f"{nome} | Shape: {df.shape}")


In [ ]:
df_pesquisa_metadados = excel_pesquisa["Metadados"]
df_pesquisa_metadados

In [ ]:
df_pesquisa = excel_pesquisa["dados_da_campanha"]
df_pesquisa

In [ ]:
df_pesquisa = df_pesquisa.set_index("id_do_respondente")
df_pesquisa

In [ ]:
full_describe(df_pesquisa)

In [ ]:
full_value_countes(df_pesquisa)

In [ ]:
# Análise do interesse no curso
print("=== ANÁLISE DO INTERESSE NO CURSO ===")

# Verificando se existe coluna de interesse
colunas_pesquisa = df_pesquisa.columns.tolist()
print(f"Colunas disponíveis: {colunas_pesquisa}")


In [ ]:
print(f"\nDistribuição de {"interesse_no_Curso"}:")
display(df_pesquisa["interesse_no_Curso"].value_counts())
print(f"Percentual de interessados: {(df_pesquisa["interesse_no_Curso"].value_counts(normalize=True) * 100).round(2)}")


In [ ]:

# Análise do preço que as pessoas pagariam
print("=== ANÁLISE DO PREÇO ===")

# Procurando por colunas relacionadas ao preço
colunas_preco = [col for col in colunas_pesquisa if "preco" in col.lower() or "valor" in col.lower() or "pagaria" in col.lower()]
print(f"Colunas de preço encontradas: {colunas_preco}")

if colunas_preco:
    for col in colunas_preco:
        print(f"\nEstatísticas de {col}:")
        print(df_pesquisa[col].describe())

        # Histograma do preço
        plt.figure(figsize=(10, 6))
        plt.hist(df_pesquisa[col].dropna(), bins=20, alpha=0.7, edgecolor="black")
        plt.title(f"Distribuição de {col}")
        plt.xlabel("Preço (R$)")
        plt.ylabel("Frequência")
        plt.grid(True, alpha=0.3)
        plt.show()

# 7.1 Análise do investimento necessário para dobrar as vendas
print("=== ANÁLISE PARA DOBRAR AS VENDAS ===")

vendas_atual = total_vendas
vendas_dobrar = vendas_atual * 2
custo_atual = total_custo

# Assumindo que a taxa de conversão se mantém
cliques_necessarios = vendas_dobrar / (tc_geral / 100)
custo_necessario = cliques_necessarios * (custo_atual / total_cliques)

print(f"Vendas atuais: {vendas_atual:,}")
print(f"Vendas desejadas: {vendas_dobrar:,}")
print(f"Cliques necessários: {cliques_necessarios:,.0f}")
print(f"Custo necessário: R$ {custo_necessario:,.2f}")
print(f"Investimento adicional necessário: R$ {custo_necessario - custo_atual:,.2f}")

# 7.2 Análise da relação entre investimento e vendas
print("=== RELAÇÃO ENTRE INVESTIMENTO E VENDAS ===")

# Correlação entre custo e vendas
correlacao_custo_vendas = df_campanha["custo"].corr(df_campanha["vendas"])
print(f"Correlação entre custo e vendas: {correlacao_custo_vendas:.3f}")

# Gráfico de dispersão
plt.figure(figsize=(10, 6))
plt.scatter(df_campanha["custo"], df_campanha["vendas"], alpha=0.6)
plt.xlabel("Custo da Campanha (R$)")
plt.ylabel("Vendas Realizadas")
plt.title("Relação entre Custo da Campanha e Vendas")
plt.grid(True, alpha=0.3)

# Linha de tendência
z = np.polyfit(df_campanha["custo"], df_campanha["vendas"], 1)
p = np.poly1d(z)
plt.plot(df_campanha["custo"], p(df_campanha["custo"]), "r--", alpha=0.8)

plt.show()

# Análise de desperdício (campanhas com baixa taxa de conversão)
print("\n=== ANÁLISE DE DESPERDÍCIO ===")
campanhas_baixa_conversao = df_campanha[df_campanha["tc"] < df_campanha["tc"].median()]
print(f"Campanhas com baixa conversão (< mediana): {len(campanhas_baixa_conversao)}")
print(f"Custo total em campanhas de baixa conversão: R$ {campanhas_baixa_conversao['custo'].sum():,.2f}")
print(f"Percentual do custo total: {(campanhas_baixa_conversao['custo'].sum() / total_custo * 100):.2f}%")

# 7.3 Análise de perfis com maior ROI
print("=== PERFIS COM MAIOR ROI ===")

# Top 10 campanhas com maior ROI
top_roi = df_campanha.nlargest(10, "roi")[["idade", "sexo", "interesses", "roi", "custo", "vendas"]]
print("Top 10 campanhas com maior ROI:")
print(top_roi.round(2))

# Análise por combinação de variáveis
if "idade" in df_campanha.columns and "sexo" in df_campanha.columns:
    print("\nROI por combinação de Idade e Sexo:")
    roi_idade_sexo = df_campanha.groupby(["idade", "sexo"])["roi"].agg(["mean", "count"]).round(2)
    roi_idade_sexo.columns = ["ROI Médio (%)", "Quantidade de Campanhas"]
    print(roi_idade_sexo.sort_values("ROI Médio (%)", ascending=False))

if "interesses" in df_campanha.columns:
    print("\nROI por Interesses:")
    roi_interesses = df_campanha.groupby("interesses")["roi"].agg(["mean", "count"]).round(2)
    roi_interesses.columns = ["ROI Médio (%)", "Quantidade de Campanhas"]
    print(roi_interesses.sort_values("ROI Médio (%)", ascending=False))

# Análise para alocação da verba
print("=== RECOMENDAÇÕES PARA ALOCAÇÃO DA VERBA ===")

VERBA_TOTAL = 20000
